File: longest-common-subpath.cc
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)
#define ROF(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (b); --i >= (a); )

const long N = 200000;
int a[N], pos[N], sa[N], r[N], h[N], x[N];

void suffix_array(int n, int m, int h[], int x[]) {
  fill_n(h, m, 0);
  REP(i, n) h[a[i]]++;
  FOR(i, 1, m) h[i] += h[i-1];
  ROF(i, 0, n) sa[--h[a[i]]] = i;
  r[sa[0]] = 0;
  FOR(i, 1, n) r[sa[i]] = r[sa[i-1]] + (a[sa[i]] != a[sa[i-1]]);
  int k = 1, l;
  for (; r[sa[n-1]] != n-1; k *= 2) {
    REP(i, n) h[r[sa[i]]] = i;
    ROF(i, 0, n) if (sa[i] >= k) x[h[r[sa[i]-k]]--] = sa[i]-k;
    FOR(i, n-k, n) x[h[r[i]]] = i;
    copy_n(x, n, sa);
    h[sa[0]] = l = 0;
    FOR(i, 1, n) h[sa[i]] = l += r[sa[i]] != r[sa[i-1]] || max(sa[i], sa[i-1]) >= n-k || r[sa[i]+k] != r[sa[i-1]+k];
    copy_n(h, n, r);
  }
  h[0] = k = 0;
  REP(i, n) {
    if (r[i]) {
      for (int j = sa[r[i]-1]; i+k < n && j+k < n && a[i+k] == a[j+k]; k++);
      h[r[i]] = k;
    }
    k && k--;
  }
}

class Solution {
public:
  int longestCommonSubpath(int n, vector<vector<int>>& paths) {
    int m = paths.size(), tot = 0;
    REP(i, m) {
      REP(j, paths[i].size()) {
        pos[tot] = i;
        a[tot++] = paths[i][j];
      }
      a[tot++] = n+i;
    }
    suffix_array(tot, n+m, h, x);

    fill_n(r, m, 0);
    int ans = 0, cnt = 0, j = 0, hd = 0, tl = 0;
    REP(i, tot-m) {
      if (!r[pos[sa[i]]]++) cnt++;
      while (hd < tl && h[x[tl-1]] > h[i]) tl--;
      x[tl++] = i;
      if (cnt < m) continue;
      while (cnt == m) {
        if (!--r[pos[sa[j]]]) cnt--;
        if (x[hd] == j++) hd++;
      }
      ans = max(ans, h[x[hd]]);
    }
    return ans;
  }
};

File: longest-common-subpath.rs
// Longest Common Subpath
impl Solution {
    pub fn longest_common_subpath(ab: i32, mut paths: Vec<Vec<i32>>) -> i32 {
        let np = paths.len();
        let mut a = vec![];
        let mut pos = vec![];
        for i in 0..np {
            a.append(&mut paths[i]);
            a.push(ab + i as i32);
            pos.resize(a.len(), i as i32);
        }

        let n = a.len();
        let m = ab as usize + np;
        let mut sa: Vec<i32> = Vec::with_capacity(n);
        let mut r: Vec<i32> = Vec::with_capacity(n);
        let mut h = vec![0; n.max(m)];
        let mut x: Vec<i32> = Vec::with_capacity(n);
        unsafe { sa.set_len(n); r.set_len(n); x.set_len(n); }
        for i in 0..n { h[a[i] as usize] += 1; }
        for i in 1..m { h[i] += h[i-1]; }
        for i in (0..n).rev() { let v = &mut h[a[i] as usize]; *v -= 1; sa[*v as usize] = i as i32; }
        r[sa[0] as usize] = 0;
        for i in 1..n { r[sa[i] as usize] = r[sa[i-1] as usize] + if a[sa[i] as usize]!=a[sa[i-1] as usize] {1} else {0}; }
        let mut k = 1;
        while r[sa[n-1] as usize] != (n-1) as i32 {
            for i in 0..n { h[r[sa[i] as usize] as usize] = i as i32;}
            for i in (0..n).rev() { if sa[i] >= k as i32 {
                let v = &mut h[r[sa[i] as usize-k] as usize]; x[*v as usize] = sa[i]-k as i32; *v -= 1; }}
            for i in n-k..n { x[h[r[i] as usize] as usize] = i as i32; }
            sa[0..n].clone_from_slice(&x[0..n]);
            h[sa[0] as usize] = 0;
            for i in 1..n { h[sa[i] as usize] = h[sa[i-1] as usize] +
                if r[sa[i] as usize]!=r[sa[i-1] as usize] || sa[i].max(sa[i-1]) as usize >= n-k ||
                    r[sa[i] as usize+k]!=r[sa[i-1] as usize+k] {1} else {0}; }
            r[0..n].clone_from_slice(&h[0..n]);
            k *= 2;
        }
        k = 0;
        h[0] = 0;
        for i in 0..n {
            if r[i] > 0 {
                let j = sa[r[i] as usize-1] as usize;
                while i.max(j)+k < n && a[i+k] == a[j+k] { k += 1; }
                h[r[i] as usize] = k as i32;
            }
            if k > 0 { k -= 1; }
        }

        for i in 0..np { r[i] = 0; }
        let mut ans = 0;
        let mut cnt = 0;
        let mut j = 0;
        let mut hd = 0;
        let mut tl = 0;
        for i in 0..n-np {
            let v = &mut r[pos[sa[i] as usize] as usize];
            if *v == 0 { cnt += 1; }
            *v += 1;
            while hd < tl && h[x[tl-1] as usize] > h[i] { tl -= 1; }
            x[tl] = i as i32;
            tl += 1;
            if cnt < np { continue; }
            while cnt == np {
                let v = &mut r[pos[sa[j] as usize] as usize];
                *v -= 1;
                if *v == 0 { cnt -= 1; }
                if x[hd] == j as i32 { hd += 1; }
                j += 1;
            }
            ans = ans.max(h[x[hd] as usize]);
        }
        ans
    }
}
